File: backend/python-project/speaking-practice-partner/api/grammar.py
"""语法检查API"""
from fastapi import HTTPException, APIRouter
from models import GrammarCheckRequest, GrammarCheckResponse, ChinglishDetectRequest, ChinglishDetectResponse, CorrectionRequest, CorrectionResult
from services.nlp import GrammarChecker, ChinglishDetector
from services.nlp.grammar_corrector import GrammarCorrector
from loguru import logger

router = APIRouter()

# 初始化服务
grammar_checker = GrammarChecker()
chinglish_detector = ChinglishDetector()
grammar_corrector = GrammarCorrector()


@router.post("/check", response_model=GrammarCheckResponse)
async def check_grammar(request: GrammarCheckRequest):
    """
    语法检查
    
    - **text**: 待检查文本
    """
    try:
        errors = grammar_checker.check_grammar(request.text)
        return GrammarCheckResponse(errors=errors)
    except Exception as e:
        logger.error(f"语法检查失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/detect-chinglish", response_model=ChinglishDetectResponse)
async def detect_chinglish(request: ChinglishDetectRequest):
    """
    中式英语检测
    
    - **text**: 待检测文本
    """
    try:
        results = chinglish_detector.detect_chinglish(request.text)
        return ChinglishDetectResponse(results=results)
    except Exception as e:
        logger.error(f"中式英语检测失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/correct", response_model=CorrectionResult)
async def correct_grammar(request: CorrectionRequest):
    """
    语法纠错（新增）
    
    使用LanguageTool + spaCy + 自定义规则进行语法纠错
    
    - **text**: 待纠错文本
    - **detailed**: 是否返回详细解释（默认True）
    """
    try:
        result = grammar_corrector.correct(request.text, request.detailed)
        return result
    except Exception as e:
        logger.error(f"语法纠错失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/utils/grammar.js
/**
 * 语法纠错客户端
 * 用于调用后端语法纠错 API
 */

import axios from 'axios'

class GrammarCorrector {
  constructor(baseUrl = null) {
    // 使用环境变量配置的 Python 后端地址
    const apiBaseUrl = baseUrl || import.meta.env.VITE_PYTHON_API_BASE_URL
    
    this.client = axios.create({
      baseURL: apiBaseUrl,
      timeout: 10000
    })
  }

  /**
   * 语法纠错
   * @param {String} text - 待纠错的文本
   * @param {Boolean} detailed - 是否返回详细解释（默认true）
   */
  async correct(text, detailed = true) {
    try {
      const response = await this.client.post('/v1/grammar/correct', {
        text,
        detailed
      })
      return response.data
    } catch (error) {
      console.error('语法纠错失败:', error)
      throw error
    }
  }

  /**
   * 语法检查
   * @param {String} text - 待检查的文本
   */
  async check(text) {
    try {
      const response = await this.client.post('/v1/grammar/check', {
        text
      })
      return response.data
    } catch (error) {
      console.error('语法检查失败:', error)
      throw error
    }
  }

  /**
   * 中式英语检测
   * @param {String} text - 待检测的文本
   */
  async detectChinglish(text) {
    try {
      const response = await this.client.post('/v1/grammar/detect-chinglish', {
        text
      })
      return response.data
    } catch (error) {
      console.error('中式英语检测失败:', error)
      throw error
    }
  }
}

/**
 * 语法纠错（便捷函数）
 * @param {String} text - 待纠错的文本
 * @param {Boolean} detailed - 是否返回详细解释
 */
export async function correctGrammar(text, detailed = true) {
  const corrector = new GrammarCorrector()
  return await corrector.correct(text, detailed)
}

export default GrammarCorrector